app/components/autocomplete/channel_mention/channel_mention.js

```diff
@@ -0,0 +1,266 @@
+        theme: PropTypes.object.isRequired,
+        actions: PropTypes.shape({
+            changePostDraft: PropTypes.func.isRequired,
+            autocompleteChannels: PropTypes.func.isRequired
+        })
+    }
+
+    static defaultProps = {
+        postDraft: ''
+    }
+
+    constructor(props) {
+        super(props);
+
+        const ds = new ListView.DataSource({
+            sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
+            rowHasChanged: (r1, r2) => r1 !== r2
+        });
+
+        this.state = {
+            active: false,
+            dataSource: ds.cloneWithRowsAndSections(props.autocompleteChannels)
+        };
+    }
+
+    componentWillReceiveProps(nextProps) {
+        const match = nextProps.postDraft.substring(0, nextProps.cursorPosition).match(CHANNEL_MENTION_REGEX);
+
+        // If not match or if user clicked on a channel
+        if (!match || this.state.mentionComplete) {
+            const nextState = {
+                active: false,
+                mentionComplete: false
+            };
+
+            // Handle the case where the user typed a ~ first and then backspaced
+            if (!nextProps.postDraft.length) {
+                nextState.matchTerm = null;
+            }
+
+            this.setState(nextState);
+            return;
+        }
+
+        const matchTerm = match[2];
+        const myChannels = nextProps.autocompleteChannels.myChannels;
+        const otherChannels = nextProps.autocompleteChannels.otherChannels;
+
+        // Show loading indicator on first pull for channels
+        if (nextProps.requestStatus === RequestStatus.STARTED && ((myChannels.length === 0 && otherChannels.length === 0) || matchTerm === '')) {
+            this.setState({
+                active: true,
+                loading: true
+            });
+            return;
+        }
+
+        // Still matching the same term that didn't return any results
+        if (match[0].startsWith(`~${this.state.matchTerm}`) && (myChannels.length === 0 && otherChannels.length === 0)) {
+            this.setState({
+                active: false
+            });
+            return;
+        }
+
+        if (matchTerm !== this.state.matchTerm) {
+            this.setState({
+                matchTerm
+            });
+
+            const {currentTeamId} = this.props;
+            this.props.actions.autocompleteChannels(currentTeamId, matchTerm);
+            return;
+        }
+
+        if (nextProps.requestStatus !== RequestStatus.STARTED && this.props.autocompleteChannels !== nextProps.autocompleteChannels) {
+            let data = {};
+            if (myChannels.length > 0) {
+                data = Object.assign({}, data, {myChannels});
+            }
+            if (otherChannels.length > 0) {
+                data = Object.assign({}, data, {otherChannels});
+            }
+
+            this.setState({
+                active: true,
+                loading: false,
```

Comment:
The component flashes when that is the only check made since we make a request on each keypress. Having the loading state done this way allows for the component to stay open and eliminate the flash.